### PR TITLE
Created functionality to direct upload video to vimeo

### LINKED ISSUE #64: Direct upload video to vimeo
When creating or editing a lesson currently we manually upload videos to vimeo and use the link while creating lessons.  This approach has certain issues

1. The process involves vimeo and instruo
2. We need to educate content creators on how this works.
3. Content team has to manually input duration at the time of creating video.
4. In case of a change of vendor from vimeo video migration is going to be very difficult.
5. In case of our own steaming setup video migration can be difficult.

TODO

1. Allow user to choose video file 
7. The videos will be uploaded to s3 bucket by default.
8. After completing file uploads a background job should fetch the uploaded file from s3 and it should push the video to vimeo directory (a preset location for course videos).
9. One a video is uploaded to vimeo get its URL and update the lessons and local contents with the vimeo URL. 
10. The video should now be playable on the platform.

## Workflow diagram

Screenshot 2024-10-22 at 1 34 00 PM

NB: No database operation is performed from the background Job. The job shall fetch and update back to the back-end application. The communication between background job and backend application is triggered by the enqueued lesson id. Other communication's can be done using internal REST API's, authentication is not required for such API's.

# Cleanups & UI

Remove the video url field, make the duration field a hidden field (duration  now should be filled automatically on choosing video. 

Show upload progress on video uploads (design is already there for progress bar)

### LINKED ISSUE #185: Fix `bin/setup` failing
Screenshot 2024-11-01 at 3 16 14 PM

### PR DESCRIPTION
**Pull request by paulsojan**
Fixes #64 
Fixes #185 

Demo: https://paulsojan.neetorecord.com/watch/c88f749d-b412-487c-8ef1-f29f9df8a523

### CODE REVIEW

**Comment by paulsojan** [comment]
@deepakkumarnd Please review

File: app/services/upload_video_to_vimeo_service.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    request.body = {
+      upload: {
+        approach: 'tus',
+        size: file.byte_size
+      },
+      name: file.filename.to_s
+    }.to_json
+
+    Net::HTTP.start(url.hostname, url.port, use_ssl: true) { |http| http.request(request) }
```
Comment (deepakkumarnd) [inline app/services/upload_video_to_vimeo_service.rb:57]: We should expect failure here and need to add exception handling when connecting to external services.

File: app/services/upload_video_to_vimeo_service.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      tempfile.rewind
+      url = URI.parse(upload_url)
+      request = Net::HTTP::Patch.new(url)
+
+      request['Tus-Resumable'] = '1.0.0'
+      request['Upload-Offset'] = '0'
+      request['Content-Type'] = 'application/offset+octet-stream'
+
+      request.body = tempfile.read
+      Net::HTTP.start(url.hostname, url.port, use_ssl: true) { |http| http.request(request) }
```
Comment (deepakkumarnd) [inline app/services/upload_video_to_vimeo_service.rb:73]: We should expect failure here and need to add exception handling when connecting to external services.

File: spec/services/upload_video_to_vimeo_service_spec.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      vimeo_video_url = local_content.video.blob.metadata['url']
+
+      expect(response).to be_a(Net::HTTPOK)
+      expect(vimeo_video_url).to eq('https://vimeo.com/1234')
+    end
+  end
+
+  private
+
+  def upload_to_vimeo_service(blob)
```
Comment (deepakkumarnd) [inline spec/services/upload_video_to_vimeo_service_spec.rb:30]: you could use subject {} to define the subject under test at the beginning of the test

subject {   UploadVideoToVimeoService.new(blob)
  end }

**Review by deepakkumarnd — CHANGES_REQUESTED** [review]

**Comment by paulsojan** [comment]
@deepakkumarnd Made changes. Please review

**Pull request merged**